src/lib/new_device/new_device.cpp

```diff
@@ -0,0 +1,351 @@
+#include <vector>
+
+#include <iomgr/iomgr.hpp>
+#include <isa-l/crc.h>
+#include <sisl/logging/logging.h>
+
+#include <homestore/homestore_decl.hpp>
+#include "new_device/new_device.h"
+#include "new_device/physical_dev.hpp"
+#include "new_device/chunk.h"
+#include "new_device/virtual_dev.hpp"
+#include "common/homestore_utils.hpp"
+#include "common/homestore_assert.hpp"
+
+namespace homestore {
+
+static int determine_open_flags(io_flag oflags) {
+    int open_flags;
+
+    switch (oflags) {
+    case io_flag::BUFFERED_IO:
+        open_flags = O_RDWR | O_CREAT;
+        break;
+    case io_flag::READ_ONLY:
+        open_flags = O_RDONLY;
+        break;
+    case io_flag::DIRECT_IO:
+        open_flags = O_RDWR | O_CREAT | O_DIRECT;
+        break;
+    default:
+        open_flags = O_RDWR | O_CREAT;
+    }
+
+    return open_flags;
+}
+
+static bool is_hdd(const std::string& devname) {
+    const iomgr::drive_type dtype = iomgr::DriveInterface::get_drive_type(devname);
+    if (dtype == iomgr::drive_type::block_hdd || dtype == iomgr::drive_type::file_on_hdd) { return true; }
+    return false;
+}
+
+static void populate_vdev_info(const vdev_parameters& vparam, uint32_t vdev_id,
+                               const std::vector< PhysicalDev* >& pdevs, vdev_info* out_info);
+
+DeviceManager::DeviceManager(const std::vector< dev_info >& devs, new_vdev_cb_t new_vdev_cb) :
+        m_dev_infos{devs}, m_new_vdev_cb{std::move(new_vdev_cb)} {
+    bool found_hdd_dev{false};
+    for (const auto& dev_info : devs) {
+        if (is_hdd(dev_info.dev_name)) {
+            HomeStoreStaticConfig::instance().hdd_drive_present = true;
+            found_hdd_dev = true;
+            LOGINFO("HDD device found: {}");
+            break;
+        }
+    }
+
+    if (found_hdd_dev) {
+        if ((HS_STATIC_CONFIG(input.data_open_flags) == io_flag::DIRECT_IO) &&
+            !HS_DYNAMIC_CONFIG(device->direct_io_mode)) {
+            // override direct i/o for HDD's
+            LOGINFO("Overridding HDD open flags from DIRECT_IO to BUFFERED_IO");
+            m_hdd_open_flags = determine_open_flags(io_flag::BUFFERED_IO);
+        } else {
+            m_hdd_open_flags = determine_open_flags(HS_STATIC_CONFIG(input.data_open_flags));
+        }
+    }
+    m_ssd_open_flags = determine_open_flags(HS_STATIC_CONFIG(input.fast_open_flags));
+
+    // Read from all the devices and check if there is a valid superblock present in those devices.
+    bool format{true};
+    for (const auto& d : devs) {
+        first_block fblk = PhysicalDev::read_first_block(d.dev_name, device_open_flags(d.dev_name));
+        if (fblk.is_valid()) {
```

Comment:
Echoing my previous comment about checking version in |is_valid", it is used to determine whether we need to format the device or not. If this is an old version (e.g. version 3, which can be some device used to be running with 1.3 Homestore stuff), we should format it, as we don't support upgrade from old version for version 4, right?

Does it make sense?

Reply:
To start with 1.3 is not a possibility because in 1.x we leave the first 4k blocks empty for a reason which is incorrect. So we will never see the first block same as this. My point is we don't need to see anything other than version 4 at the moment. When we have next major version, we handle that time.